bazel_repositories_priv

```diff
@@ -0,0 +1,292 @@
+        else:
+            prefix = url_path.parts[2]
+        filename = prefix + "-" + filename
+    download(url, filename)
+
+    log.info(f"Checking sha256 sum of {filename}")
+    sha256sum = hashlib.sha256()
+    with open(filename, "rb") as f:
+        while True:
+            buf = f.read(65536)
+            if not buf:
+                break
+            sha256sum.update(buf)
+    sha256sum_digest = sha256sum.hexdigest()
+
+    log.info(f"Storing {filename}")
+    os.makedirs(os.path.join(CACHEDIR, sha256sum_digest), exist_ok=True)
+    cache_filename = os.path.join(CACHEDIR, sha256sum_digest, "file")
+    shutil.move(filename, cache_filename)
+
+    return cache_filename
+
+
+def process_urls(urls: typing.List[str],
+                 exclude: typing.List[str]) -> typing.Tuple[
+                     typing.List[str], typing.List[str]]:
+    """Download and store archives from the given list of URLs. Return the list
+    of filtered URLs and the list of filtered out cache files which need to be
+    removed later.
+    """
+    archives = set()
+    filtered_urls = []
+    excluded = []
+    # Bazel configurations may specify multiple URLs per one dependency. Here
+    # we download each dependency only once and try next URLs only if the
+    # download from previous URL was not possible.
+    for url in urls:
+        append_url = True
+        url_path = pathlib.Path(urllib_parse.urlparse(url).path)
+        filename = url_path.parts[-1]
+        if filename in archives:
+            continue
+        try:
+            cache_filename = process_url(url, url_path, filename)
+        except (urllib_error.HTTPError, urllib_error.URLError):
+            log.warning(f"could not download {url}")
+            continue
+        else:
+            archives.add(filename)
+            # If dependency is excluded, don't add it to the result list,
+            # remove the archive from the working dir, but keep it in the
+            # cache (otherwise next iterations of ´bazel cache` will not work).
+            for e in exclude:
+                if e in url:
+                    log.info(f"excluding dependency {url}")
+                    append_url = False
+            if append_url:
+                filtered_urls.append(url)
+            else:
+                excluded.append(cache_filename)
+    return filtered_urls, excluded
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description=description,
+        formatter_class=argparse.RawDescriptionHelpFormatter)
+    parser.add_argument("--archive")
+    parser.add_argument("--bazel-args")
+    parser.add_argument("--exclude")
+    parser.add_argument("--home")
+    parser.add_argument("--target")
+    parser.add_argument("--user")
+    args = parser.parse_args()
+
+    log.debug(f"target: {args.target}")
+
+    exclude = args.exclude.split(",") if args.exclude else []
+    root_dir = find_rootdir(args.archive)
+
+    if os.getuid() != 0:
+        exit("You have to be root")
```

Comment:
I'm not a big fan of checking for UID=0.  Root isn't always privileged, and it's possible to use filesystem capabilities or similar mechanisms to run privileged without the specific user.  Ideally, one would use [prctl](https://pythonhosted.org/python-prctl/) (or maybe [libcap-ng](https://github.com/stevegrubb/libcap-ng), which I'm less familiar with) and check for the specific capabilities needed.

Reply:
I will try prctl or libcap-ng.